### PR TITLE
Company is "invisible" if no members

### PR DESCRIPTION
**Pull request by patmbolger**
PT Story: https://www.pivotaltracker.com/story/show/153496033

Changes proposed in this pull request:

1. Logic changes required to make app compliant with rule #4 (https://github.com/AgileVentures/shf-project/wiki/Business-rules-around-payments-(membership-and-H-Branding)

Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

File: app/controllers/companies_controller.rb [hunk 1]
```diff
@@ -19,7 +19,9 @@ def index
                           .includes(addresses: [ :region, :kommun ])
                           .joins(addresses: [ :region, :kommun ])
 
-    @all_companies = @all_companies.branding_licensed unless current_user.admin?
+    unless current_user.admin?
+      @all_companies = @all_companies.branding_licensed.with_members
+    end
 
     # The last qualifier ("joins") on above statement ("addresses: :region") is
```
Comment (weedySeaDragon) [inline app/controllers/companies_controller.rb]: need to either move this comment up (before the `unless....` conditional) or move the conditional down below this comment.

File: features/step_definitions/company_steps.rb [hunk 2]
```diff
@@ -13,6 +13,22 @@
...
 end
 
+And(/^the following simple companies exist:$/) do |table|
+  table.hashes.each do |company|
+    region = company.delete('region') || 'Stockholm'
+    kommun = company.delete('kommun') || 'Stockholm'
+    visibility = company.delete('visibility') || 'street_address'
+
+    cmpy = FactoryGirl.build(:company, company)
+    cmpy.save(validate: false)
```
Comment (weedySeaDragon) [inline features/step_definitions/company_steps.rb]: It looks like you're doing this to get around the company number validation, yes?
But then you're running an integration test with _invalid data_ and expecting it to pass.  It shouldn't.

(Also, someone might come along later and use this background step, make a change to a company and then be surprised by a validation issue. But this is secondary; invalid data in an integration test should not pass [unless by "pass" you mean 'flag the data as invalid'])

File: features/step_definitions/membership_application_steps.rb [hunk 3]
```diff
@@ -27,10 +28,27 @@
 
 And(/^the following simple applications exist:$/) do |table|
```
Comment (weedySeaDragon) [inline features/step_definitions/membership_application_steps.rb]: Ditto much of my comment about the step name for ".... simple companies exist".

In this case it's not so clear if you're putting in invalid data (which is kind of a problem in itself).

Just as importantly: the step name (text / regex) is not communicating the _intent_. 
What do you mean by a "simple" application?  What is it you're trying to accomplish with this? (I don't get it)

Someone might come along later and use this background step, make a change to an application and then be surprised by a validation issue.

File: features/step_definitions/membership_application_steps.rb [hunk 4]
```diff
@@ -8,13 +8,14 @@
...
    end
-   contact_email = hash['contact_email'] && ! hash['contact_email'].empty? ? 
+   contact_email = hash['contact_email'] && ! hash['contact_email'].empty? ?
                    hash['contact_email'] : hash[:user_email]
+
    ma = FactoryGirl.create(:membership_application,
                             attributes.merge(user: user,
                             company: company,
                             contact_email: contact_email))
-   ma.state = hash['state'].to_sym if hash.has_key?('state')
```
Comment (weedySeaDragon) [inline features/step_definitions/membership_application_steps.rb]: Why did you remove this?  How will we create an application in a particular state?
(It's quite likely that we _can_ without this and I just don't understand how.)

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
You've created the "... simple companies exist" and "... simple applications exist" for some reason(s) but it's not clear to me why.  I don't doubt you have good reasons for wanting them, but I believe that they way you've done it is problematic.

File: features/step_definitions/membership_application_steps.rb (hunk 4 above)
Comment (patmbolger) [reply features/step_definitions/membership_application_steps.rb]: It was unneeded in that `state`, in the hash table, does not have to be converted to a symbol in order to be assigned to the application `state` attribute.   So, the hash value for `state` was already saved when the application was created.

(It was, in fact, also ineffective, as the application (`ma`) was never saved again and thus this assignment was useless.)

File: features/step_definitions/company_steps.rb (hunk 2 above)
Comment (patmbolger) [reply features/step_definitions/company_steps.rb]: I am doing this for tests that (for most part) require many instances (or, at least, certainly more than a few).  These are things like pagination (where I need at least 21+ instances), items-per-page in paginated table (ditto).

(the steps are used in `view_membership_applications.feature` and `view_all_companies.feature`).

The validity of the attribute `company_number` (for `Company` _or_ `MembershipApplication`) is not important in these tests.

And, setting up 21+ instances in the background (where the same company_number has to be used for companies, applications, payments) is very laborious if we have to use valid company_number's.  And it increases the opportunity to make mistakes as well (in setting up the data).

So, I chose this way to set up and maintain the tests.

I think of this as akin to stubbing or mocking code features that are not needed for specific tests, and get in the way of isolating specific code aspects under test.  (Also similar to the question of mocking external services calls for acceptance tests (some projects do, some don't)).  (and, yes, I realize this is rather _weak_ rationalization for this particular case)

I do agree that we need to be careful with this approach for _acceptance_ tests.  

I could certainly change it if needed.  We could do one or more of the following:

1) Change the step method names (and add comments) making the intended usage of the steps clear.
2) Continue to use fake numbers, but in the step methods stub the validation method in each model.
3) Change back to using valid org numbers.
4) Other?

Comment (weedySeaDragon) [reply features/step_definitions/company_steps.rb]: I'm going to write things down really explicitly to help me organize my thoughts about this and so that we can break this down and figure out where to make a change (or not).  I also want to be explicit so that those that aren't familiar with these testing concepts can follow along. I also am a little pressed for time, so can't go back and edit this too much.  I know it's repetitive in places.  Bear with me.  (And I'm numbering my points just so it's easier to refer back to them.  I'm sorry it looks so pretentious. That's not my intent.)

Here's a quote from Sandi Metz's "Practical Object-Oriented Design in Ruby: An Agile Primer" that I think points exactly to what we're talking about:

> Injecting the same objects at run-time ensures that tests break correctly but may lead to long running tests.  Alternatively, injecting doubles can speed up tests but leave them vulnerable to constructing a fantasy world where tests work but the application fails.
(p. 213)

(She's talking specifically about "roles," "dependency injection," and "doubles, " but the crux holds.)

1. This is an integration test (BDD), so it's "outside-in." (vs. a unit test, (TDD), which is "inside-out.")  The goals of an integration test are:
   a. test interaction of system components
   b. express (document) a feature of the system
   c. ensure that this feature/test is not coupled to specific testing conditions/set-up
      - else the testing conditions do not reflect the 'real world' conditions 
   d.  ensure that the system implementation can change without requiring the test to change  (Thus you're really testing the 'API' of the system components and not their implementation.)
_(There are likely others; this is what my brain produces ATM)_

2. What is the focus of this _Feature?_  (We don't have that written for this test.  ARGH!)  Here's my take on what this feature is:

   **Feature:  Show all 'complete' companies to visitors, users, and members in their chosen language and allow viewers to control how many companies are shown per page.**

   - Obviously we need to first all agree on this.
   - is this feature covering too much? probably.  It seems likely that we should pull out some of the scenarios into their own feature.  But could changing the language while using pagination uncover some UI bug/issue?  Perhaps it could.  But maybe that means that it a different feature:  "A viewer should be able to change the language while also changing the number of items that are shown per page."

3. Using invalid data means that this Feature (and others that use the `..simple companies...` step) are *coupled to this expression (structure + representation) of the data.*  That makes the tests both deviate from the real system and makes them rigid.
  > setting up 21+ instances in the background (where the same company_number has to be used for companies, applications, payments) is very laborious if we have to use valid company_number's. And it increases the opportunity to make mistakes as well (in setting up the data)
   - Setting up real-world data is often hard.  Doing so correctly is what makes test valuable.  If your tests pass with faked data that would not be considered valid in the real world with your system, what value is your test providing?  (Katrina Owens, co-author with Sandi Metz of the "99 Bottles..." book) has a presentation titled ["467 tests, 0 failures, 0 confidence"](https://speakerdeck.com/railsberry/zero-confidence-by-katrina-owen)
    - The data should not change that much so once it's set up, the work is done.  There should be very little change done to the data, so maintenance costs are low and the risk of introducing mistakes is slim.  If the test data is changing, then either the real world is highly in flux, or the data is (inappropriately) highly coupled to the test.
    - We're up to it.  "21+" is a pretty small number to deal with.  And computers are super-good at tackling laborious, repetitive work based on well-defined patterns. ;-)  


4.   I have been thinking about it as an **"acceptance test"**, and to me that means exercising every component/system involved in a real way so we can uncover any problems with the interactions and side-effects.  I tend to see the word "Feature" as a synonym for "acceptance test."  But we have not been clear with this project as to whether that's actually the case. 
   So I wonder if you're seeing this as a **"functional integration test,"** which I define of as "something that proves that the system works correctly _for a specific functional area._"  It definitely exercises/test interactions between specific system components (which may or may not happen to include external systems).  It does not purport to exercise components that are not directly involved with the specific functional area.
   There is definitely overlap between those 2 types of tests (as there should be).

5. So if this is an **acceptance test** then we should _not_ be using fake anything.  _But_ if this is a **functional integration test** then we can use some 'fake' things that are absolutely not being tested if and only if (iff) those fake things (a) are not being tested, _and_ (b) the side-effects (behavior) of those things do not affect the things/area being tested.  (b) is the tricky one to guarantee.


6.  Part of what is happening is that our system is interacting an _external system:_  the **orgnummber gem.**  This is what restricts us to using only "valid" company numbers (a.k.a. org nummers / orgNrs).
  - This system boundary is where we can swap in something:  we can put in a "fake orgnummer gem."  This "fake orgnummber gem" might accept _any_ number as valid.  

7.  What are the consequences of using a fake component in an integration test? (Using a "fake orgnummber gem.")   It is clearly different from using a fake component (mock, stub, double) in a unit test because the focus of a unit test is so different.
The consequences depend on the scope and purpose/intent of this feature/test.

Whew.

So:  Here's where I end up after all of that:

1. We should agree on and write down the different types of tests we're writing.  "unit" (typically our RSpec tests), "functional integration,"  "acceptance,"  and whatever else we decide.
3. The intent of this is a "functional integration test"
   - that intent must be made clear in the Feature title and in the description of the feature.
4. Because this is a "functional integration test" we can 'fake' some components that are not under test
5. We should fake the `orgnummber` gem.  We are thus going to substitute in a 'fake' external system that fufills some _role._  (Doing so correctly means we could swap in/out any component/code that fufills that role as long as it responds to the public interface/API.)
6. Change the step to read `Given the following companies exist without caring if their company number is valid:` so that the assumptions are obvious should someone want to use this step later.
7. Use **fake company number**  instead of 'company_number' to make it clear that this data is different from what is used in other Features.  Ex:
```
    Given the following companies exist without caring if their company number is valid:
      | name      | fake_company_number  | email           | region       | kommun  |

```
8. Create a trait in the Company Factory that will create (or build) a one of these companies.  Let the _factory_ handle using a 'fake' orgnummer (this is where you use a _double_).
  - this line in the implementation of the `And(/^the following simple companies exist:$/) do |table|` step can lead to false positives because some other attribute might be saved in an invalid state and _we'd never know_
```
cmpy.save(validate: false)
```
This line **mus**t changed to just a 'normal' `save` to avoid false positives with this step and any test that uses it

9. using the 'fake' `orgnummer` double means that you will be stubbing the validation process for a `Company` since that is where the interaction happens between a `Company` and `orgnummer` (the external system).

---

That's what I have time for today, @patmbolger .  I'm really interested in thoughts and feedback from @AgileVentures/shf-project-team  and all.

**Comment by patmbolger** [comment]
I removed the "simple" applications and companies and reverted back to using real org numbers.

I altered application and payment fixture step methods to allow specification of either company number _or_ company name - this greatly reduces complexity in setting up test data that involves all three entities.

(I think at some point we could provide the ability to specify that an org number should be created on-the-fly (perhaps enter "generate" in the fixture table column).  For that, we could extract `get_company_number` from `seed_helpers.rb` and make that available to tests as well.)

File: spec/models/company_spec.rb [hunk 5]
```diff
@@ -429,4 +429,51 @@
...
   end
+
+  describe 'scope: with_members' do
+    let(:cmpy1) { create(:company, company_number: '5560360793') }
+    let(:cmpy2) { create(:company, company_number: '5562252998') }
+
+    let(:user1) { create(:user) }
+    let(:user2) { create(:user) }
+
+    let(:app1_user1) do
```
Comment (weedySeaDragon) [inline spec/models/company_spec.rb]: can these be renamed to `app_co1_user1` `app_co1_user2` and `app_co2_user2` ?

 I was (easily) confused:  the naming led me to assume that 1 application had 2 users.  Inserting "co" for 'company' would really improve readability and make maintenance easier.

File: spec/models/company_spec.rb [hunk 6]
```diff
@@ -429,4 +429,51 @@
...
+      expect(Company.with_members).to be_empty
+    end
+
+    it 'returns all companies with members' do
+      app1_user1.start_review; app1_user1.accept!
+      user1.update(member: true)
+
+      expect(Company.with_members).to contain_exactly(cmpy1)
+
+      app2.start_review; app2.accept!
```
Comment (weedySeaDragon) [inline spec/models/company_spec.rb]: use a separate line
(2 ruby statements on 1 line is not a style we've adopted)

File: spec/models/company_spec.rb [hunk 7]
```diff
@@ -429,4 +429,51 @@
...
+      user2.update(member: true)
+
+      expect(Company.with_members).to contain_exactly(cmpy1, cmpy2)
+    end
+
+    it 'returns company only once even if multiple members' do
+      app1_user1.start_review; app1_user1.accept!
```
Comment (weedySeaDragon) [inline spec/models/company_spec.rb]: use a separate line
(2 ruby statements on 1 line is not a style we've adopted)

File: spec/models/company_spec.rb [hunk 8]
```diff
@@ -429,4 +429,51 @@
...
+    end
+
+    before(:each) { app1_user1; app1_user2; app2 }
+
+    it 'returns no companies if no members' do
+      expect(Company.with_members).to be_empty
+    end
+
+    it 'returns all companies with members' do
+      app1_user1.start_review; app1_user1.accept!
```
Comment (weedySeaDragon) [inline spec/models/company_spec.rb]: use a separate line
(2 ruby statements on 1 line is not a style we've adopted)

File: features/view_all_companies.feature [hunk 9]
```diff
@@ -17,100 +17,106 @@ Feature: As a visitor,
...
+      | Company20 | 0000000020     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company21 | 0000000021     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company22 | 0000000022     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company23 | 0000000023     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company24 | 0000000024     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company25 | 0000000025     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company26 | 0000000026     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company27 | 0000000027     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company28 | 0000000028     | cmpy13@mail.com | Stockholm    | Alingsås|
+      | Company29 | 0000000029     | cmpy13@mail.com | Stockholm    | Alingsås|
```
Comment (weedySeaDragon) [inline features/view_all_companies.feature]: Can we rename `Company29`  to `NoBrandingFeePaid` ?  (or something that conveys that meaning)  That would make the purpose and set-up/data of this company very clear.  I had to look through _all_ of the data to confirm that is what this company indeed testing.

Likewise, can `Company28` be renamed to something more meaningful?  I see that there has been a branding fee paid, but I don't know how this quite satisfies the "no members" condition that you're testing in `Scenario: Companies lacking branding payment or members not shown`

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
We need a Scenario that covers `Admin can view all companies including those with no members or   no branding fee paid` 

It would cover this condition in  CompaniesController#index:

```
    unless current_user.admin?
      @all_companies = @all_companies.branding_licensed.with_members
    end
```
Not sure if we want to put it into `view_all_companies.feature` or not.  (I don't care)

Other than that, my comments are largely about style.

**Comment by patmbolger** [comment]
@weedySeaDragon - changes made ... thanks for the comments.

**Review by weedySeaDragon — APPROVED** [review]

**Review by RobertCram — APPROVED** [review]
